azure-kusto-ingest/source/ingestionProperties.ts

```diff
@@ -52,28 +61,211 @@ export enum ReportMethod {
     Queue = 0
 }
 
-class ColumnMapping {
+export enum FieldTransformation {
+    PropertyBagArrayToDictionary = "PropertyBagArrayToDictionary",
+    DateTimeFromUnixSeconds = "DateTimeFromUnixSeconds",
+    DateTimeFromUnixMilliseconds = "DateTimeFromUnixMilliseconds",
+    DateTimeFromUnixMicroseconds = "DateTimeFromUnixMicroseconds",
+    DateTimeFromUnixNanoseconds = "DateTimeFromUnixNanoseconds",
+}
+
+export enum ConstantTransformation {
+    SourceLocation = "SourceLocation",
+    SourceLineNumber = "SourceLineNumber",
+}
+
+export type Transformation = FieldTransformation | ConstantTransformation;
+
+interface MappingProperties {
+    Field?: string;
+    Path?: string;
+    Ordinal?: number;
+    ConstValue?: string;
+    Transform?: Transformation;
+}
+
+type MappingPropertiesStrings = {
+    [key in keyof MappingProperties]: string
+}
+
+interface ApiColumnMapping {
+    Column: string,
+    DataType?: string,
+    Properties?: MappingPropertiesStrings
+}
+
+abstract class ColumnMapping {
+    constructor(readonly columnName: string, readonly cslDataType?: string, readonly Properties?: MappingProperties) {
+    }
+
+    public abstract mappingType(): IngestionMappingType;
+
+    public toApiMapping(): ApiColumnMapping {
+        const result: ApiColumnMapping = {
+            Column: this.columnName,
+        }
+        if (this.cslDataType) {
+            result.DataType = this.cslDataType;
+        }
+
+        if (this.Properties) {
+            result.Properties = {};
+            for (const key in this.Properties) {
+                if (this.Properties.hasOwnProperty(key)) {
+                    const typedKey = key as keyof MappingProperties;
+                    const property = this.Properties[typedKey];
+                    if (property !== undefined && property !== null) {
+                        result.Properties[typedKey] = property.toString();
+                    }
+                }
+            }
+        }
+        return result;
+    }
 }
 
 export class CsvColumnMapping extends ColumnMapping {
-    constructor(readonly columnName: string, readonly cslDataType: string, readonly ordinal: string) {
-        super();
+    /**
+     * @deprecated Use the factory methods instead.
+     */
+    protected constructor(readonly columnName: string, readonly cslDataType?: string, readonly ordinal?: string, constantValue?: string) {
+        super(columnName, cslDataType, { Ordinal: ordinal === undefined ? undefined : parseInt(ordinal, 10), ConstValue: constantValue });
+    }
+
+    public static withOrdinal(columnName: string, ordinal: number, cslDataType?: string): CsvColumnMapping {
```

Comment:
It's not a normal builder pattern here, since all of the ctors are mutually exclusive. 
You can't have both Path and Field, and these ctors prevent you from doing so accidentally.

Reply:
Discussed with team - keeping this approach